### PR TITLE
MONGOID-5274 / MONGOID-5142 - Rework #touch with embedded documents

### PR DESCRIPTION
**Pull request by johnnyshields**
Fixes MONGOID-5142
Fixes MONGOID-5274
Correct fix for MONGOID-5136 (the previous one was non-optimal)
Replaces PR https://github.com/mongodb/mongoid/pull/5039 (includes specs from it, which are passing here)

## Changes

- Perform all touch operations on the document in a single write operation (MONGOID-5142)
- Remove touches from the atomic set operations, so that they are no longer there when saving the next time.
- Cleanup/refactor touch method to be much more readable. The code is also now symmetrical for parents and child docs, everything is nicely recursive.

### CODE REVIEW

**Comment by johnnyshields** [comment]
@p-mongo if you can help with the specs here would be greatly appreciated.

**Comment by p-mongo** [comment]
Hi @johnnyshields , can you please update the title and description of this PR to reflect what it contains? MONGOID-5136 is already fixed. 

Can you also clarify whether this PR is ready for review? If it isn't it can be marked [wip] or draft.

**Comment by johnnyshields** [comment]
@p-mongo I've marked this as WIP. The ticket is MONGOID-5142. I was hoping someone from the Mongo team could contribute the specs listed as TODOs.

**Comment by johnnyshields** [comment]
I've raised https://github.com/mongodb/mongoid/pull/5202 to point out some strange/incorrect behavior re: `#touch` method. This PR is on hold until the issues are resolved.

**Comment by johnnyshields** [comment]
@Neilshweky please review this.

File: lib/mongoid/touchable.rb
Comment (Neilshweky) [inline lib/mongoid/touchable.rb:82]: TIL this is `reduce` without having to explicitly return the object. Cool.

**Comment by Neilshweky** [comment]
If I understand this correctly, your PR makes the following changes: 
- touches are now persisted immediately (backwards breaking).
  - which causes touch callbacks to be called earlier.
- touches are no longer included in the atomic_updates calculation on `save`.
  - can you put a test for this?
- touch operations are done in one operation.
- embedded touches no longer use after_touch callbacks to update associated documents.

Would you say this is a good summary of the effective changes in this PR?

I think the only backwards breaking change here would be the fact that touches are persisted immediately. Personally this doesn't bother me as a change, although release notes would have to be written about it. @p-mongo is that change okay in your opinion?

**Review by Neilshweky — COMMENTED** [review]
Please review my comments above.

File: lib/mongoid/association/options.rb
Comment (Neilshweky) [inline lib/mongoid/association/options.rb:118]: # @return [ true | false ] returns true if this association is
#   automatically touched, false otherwise. The default is false.

**Comment by johnnyshields** [comment]
@Neilshweky thanks for the review. Let me first respond to your summary of changes so we're on same page:

> 1. touches are now persisted immediately (backwards breaking). which causes touch callbacks to be called earlier.
I think the old code was the same. The old code has `_root.collection.find(selector).update_one(...)`. Please double check this one?

> 2. touches are no longer included in the atomic_updates calculation on save. can you put a test for this?
No. `touch_atomic_updates` was never actually part of `atomic_updates`. Its code just happened to be in the same file, where it doesn't belong.

> 3. touch operations are done in one operation.
Yes. This is the primary intention of the PR.

> 4. embedded touches no longer use after_touch callbacks to update associated documents.
In the new code, calling the `#touch` method on an embedded now recursively touches its parent(s) within the `#touch` method itself, so there's no need for to use `after_touch` callbacks. So this is not actually a change to the end result behavior.

File: lib/mongoid/touchable.rb
Comment (johnnyshields) [reply lib/mongoid/touchable.rb:82]: I learned this and many other lessons from Rubocop 😅

File: lib/mongoid/touchable.rb
Comment (johnnyshields) [reply lib/mongoid/touchable.rb:76]: Hmm... I'm keeping this consistent with the old `set_field_atomic_updates` method. Calling `atomic_updates` will build the updates, and I think `$set` will then include both `atomic_sets` and `delayed_atomic_sets`. If you're 100% confident there won't be knock-on effects, I'm fine to change, otherwise I'd say do it in a follow-up PR.

File: lib/mongoid/association/options.rb
Comment (johnnyshields) [reply lib/mongoid/association/options.rb]: Done

File: spec/mongoid/touchable_spec.rb
Comment (johnnyshields) [reply spec/mongoid/touchable_spec.rb:105]: I've modified it. Erm... its a bit strange here, ":touch is not set" is for embedded while ":touch is false" is for referenced; the specs themselves are duplicated code. I've improved it now with this:

```ruby
      shared_examples 'does not update the parent when :touch is not set' do
        it_behaves_like 'does not update the parent when :touch is false'
     end
```

Problem solved.

**Comment by p-mongo** [comment]
Neither https://jira.mongodb.org/browse/MONGOID-5142 nor https://jira.mongodb.org/browse/MONGOID-5142 specify making touch perform an immediate write, so I would say this shouldn't be done.

If you believe it is a desired behavior, @johnnyshields please feel free to create a ticket for it describing your use case for it. My thinking is that perhaps `touch!` could be added which would immediately persist the write (and that would have to deal with validations, for example what if there is a validation requiring updated_at to be older than a certain time, touching such a document could make it invalid).

**Comment by p-mongo** [comment]
https://jira.mongodb.org/browse/MONGOID-5274 was the other ticket.

**Comment by johnnyshields** [comment]
> perform an immediate write, so I would say this shouldn't be done.

@p-mongo correct me if I'm wrong, but as mentioned in comment above to Neil it seems the existing code also performs an immediate write. See `_root.collection.find(selector).update_one(...)` in legacy code.

**Comment by johnnyshields** [comment]
@Neilshweky I've responded to all your comments here. Please kindly resolve the comments or give further response.

**Comment by Neilshweky** [comment]
> > perform an immediate write, so I would say this shouldn't be done.
> 
> @p-mongo as mentioned in comment above to Neil, I believe the existing code also performs an immediate write. See `_root.collection.find(selector).update_one(...)` in existing code.
> 
> I agree that an immediate write is **not** desirable, however, here I intentionally tried to keep the behavior consistent. I think we should defer the write in a follow-up ticket.

I can confirm that the current functionality on master is to persist touch operations immediately.

**Comment by Neilshweky** [comment]
> Can you add a test for touching with grandchildren? I would like to see how this works with multiple generations of descendants.

@johnnyshields can you take care of this and then I'll give this another look?

File: spec/mongoid/touchable_spec.rb
Comment (Neilshweky) [reply spec/mongoid/touchable_spec.rb:105]: Do you think you can make the models consistent in touchable_spec? 

Right now we have:

- TouchableSpec
  - Embedded
    - Entrance (touch not set on embedded_in)
    - Floor (touch true on embedded_in)
  - Referenced
    - Entrance (touch false on embedded_in)
    - Floor (touch true on embedded_in)

Can you change entrance on embedded to have touch: false for Embedded, and create a third model (maybe exits?) that has touch not set, then explicitly create tests for touch not being set?

**Review by Neilshweky — COMMENTED** [review]
Please make the testing changes mentioned above:
- multi-generational test
- third model with touch not set

**Comment by johnnyshields** [comment]
@Neilshweky thanks for the fast response.

> Please make the testing changes mentioned above

Sure I'll do in next 1-2 days

> I can confirm that the current functionality on master is to persist touch operations immediately.

I've raised https://jira.mongodb.org/browse/MONGOID-5392 to track the fix for this. It should be done separately from this PR.

**Comment by johnnyshields** [comment]
@Neilshweky so firstly great idea on doing multi-level specs. I've committed some WIP, however, I found a bug which is a blocker: https://jira.mongodb.org/browse/MONGOID-5398 - Classnames with modules dont work for `embedded_model._association&.inverse_association`. Refer to ticket for example.

(Given this bug, I'm sort of confused on how any specs are working at all with embeds here. Strange.)

**Comment by johnnyshields** [comment]
@Neilshweky polite reminder -- I'm still blocked here due to MONGOID-5398. I'm glad to resume work and complete this PR whenever MONGOID-5398 is resolved.

**Comment by Neilshweky** [comment]
Hey @johnnyshields, taking another look at this rn.

A few things:
1. I checked out this PR and ran the tests in touchable_spec and they all seem to be failing. Why are they failing?
2. Do you have a test that shows the failure related to MONGOID-5398? Can you put one in touchable_spec and mark it pending?
3. MONGOID-5398 seems to be a little far out, and I think it might require a major version (just conjecture but that's what I would guess), so I'm wondering if there's a way we can get around this. I have some ideas so I would love for you to get those tests fixed and put in that test for the module problems.

**Comment by johnnyshields** [comment]
@Neilshweky

1. I encountered MONGOID-5398 while writing the tests, so I just committed messy WIP here; I'm aware the tests aren't passing.

2. MONGOID-5398 contains a clear demonstration of the failure. The specs are the ones at the end of `touchable_spec.rb` with the parent/child/grandchild touches. If you read the ticket body of MONGOID-5398 it should be obvious how it correlates to what I have in `touchable_spec.rb`.

3. / 4. The workaround would be to change the tests classes here to get rid of namespaces, and that would sidestep MONGOID-5398 and make the PR tests **appear to be green**. **However** the touch functionality would still be **broken in the real-world** for embedded classes with namespaces, which is a very common use case (e.g. 100% of embedded classes in my app are), so I think this is not a particularly meaningful approach.

**Comment by Neilshweky** [comment]
@johnnyshields I personally think a lot of what's in this PR is good and I would like for some of these changes to make it into master...

Would you mind cleaning up this PR and getting it into a form that would be releasable other than the MONGOID-5398 problems? 

I am generally okay with putting this into master and just saying that classes inside modules need to include a class_name option on their associations in the context of this ticket since that doesn't really have anything to do with this ticket, it just is other buggy behavior.

I also wonder if there's a way to avoid using the `inverse_association` method altogether so I would like to play with it if you don't mind.

**Comment by johnnyshields** [comment]
OK, I was unaware that adding `class_name` fixes [MONGOID-5398](https://jira.mongodb.org/browse/MONGOID-5398). I've just tested and it indeed works 🎉. That's enough for me to get around the blocker here and complete this PR. Please give me 2-3 weeks or so b/c I'm a bit busy at [work](https://www.tablecheck.com/en/join/) recently.

**Comment by Neilshweky** [comment]
> OK, I was unaware that adding `class_name` fixes [MONGOID-5398](https://jira.mongodb.org/browse/MONGOID-5398). I've just tested and it indeed works 🎉. That's enough for me to get around the blocker here and complete this PR. Please give me 2-3 weeks or so b/c I'm a bit busy at [work](https://www.tablecheck.com/en/join/) recently.

Hey @johnnyshields, just checking on this!

**Comment by Neilshweky** [comment]
Hey @johnnyshields, two things: 
- I have started reviewing some of those PRs:  
  - I reviewed #5267 and I'm waiting on your answer
  - Those two doc PRs I plan on reading those before the end of the week
- I believe that this PR could add value that we would be ready to use in the current state of Mongoid. Some of the other PRs that you have open are part of larger projects (i.e. attribute storage refactoring), that we simply have not started. 

Now, in order for us to accept this PR we need a few things:
- Can you type up an example with two levels of embedded documents and type up the queries before and after these changes?
  - This would demonstrate the value of this PR, as well as the correctness of the implementation. 
- Can you fix all of the tests that are still broken?

**Comment by johnnyshields** [comment]
@Neilshweky thank you for your help to get the other PRs reviewed. I will try to do a pass on this PR over the weekend.

**Comment by johnnyshields** [comment]
@Neilshweky need a bit more time on this but definitely still in my queue.

**Comment by Neilshweky** [comment]
> @Neilshweky need a bit more time on this but definitely still in my queue.

Hey what's the status on this?

**Comment by johnnyshields** [comment]
@Neilshweky I'll look at it now.

**Comment by johnnyshields** [comment]
@Neilshweky all multi-level specs done; this is ready for final review. Good idea by the way, I did catch one edge case while writing the multi-level specs.

**Comment by Neilshweky** [comment]
@johnnyshields looks like there's still some failures

**Comment by johnnyshields** [comment]
@Neilshweky please check my [most recent commit](https://github.com/mongodb/mongoid/pull/5045/commits/e9b1d38c3bdae806db24970f03208ebef741f71f). I believe the failing spec was incorrect; multi-level touch should not work unless `touch: true` is explicitly set on the relations.

**Comment by johnnyshields** [comment]
Tests are green 🎉

File: spec/mongoid/touchable_spec.rb
Comment (johnnyshields) [reply spec/mongoid/touchable_spec.rb:105]: Adding a third model can be done as part of [MONGOID-5016](https://jira.mongodb.org/browse/MONGOID-5016) where we plan to change the default behavior.

File: spec/mongoid/touchable_spec.rb
Comment (Neilshweky) [inline spec/mongoid/touchable_spec.rb:1005]: is there a reason you didn't expand these tests into `child touch: true/false?` like you did the other ones?

File: lib/mongoid/touchable.rb
Comment (Neilshweky) [inline lib/mongoid/touchable.rb:85]: maybe make this as a set being that we're checking membership in it? Marginal performance improvement but no reason not to...

File: lib/mongoid/touchable.rb
Comment (Neilshweky) [inline lib/mongoid/touchable.rb:49]: So I'm tracing through this code and this feels a little repetitive, can you explain why this is necessary? I'll explain... 

At the deepest child (take the building > floor > sofa example), the `__gather_touch_updates` call returns:
```
{ "floors.0.sofas.0.updated_at" => 2022-10-03 15:12:53 UTC }
```
Now one level up from that (the floor level): 
```
{ "floors.0.updated_at" => 2022-10-03 15:12:53 UTC, "floors.0.sofas.0.updated_at" => 2022-10-03 15:12:53 UTC }
```
And at the top (building) level: 
```
{ "updated_at" => 2022-10-03 15:12:53 UTC, "floors.0.updated_at" => 2022-10-03 15:12:53 UTC, "floors.0.sofas.0.updated_at" => 2022-10-03 15:12:53 UTC }
```
This, I think, results from the fact that the atomic updates of a child are included in its parent (does this hold always?) 

Would a more efficient solution be to find the greatest touchable ancestor and ``__extract_touches_from_atomic_sets`` once from there?

File: lib/mongoid/touchable.rb
Comment (Neilshweky) [inline lib/mongoid/touchable.rb:150]: I have a feeling this conditional is unnecessary... If the touch_field doesn't exist we'll just pass `nil` anyway which is the default parameter value... Might not be worth changing but just a note.

**Review by Neilshweky — COMMENTED** [review]
@johnnyshields great work here! I have a few comments below if you don't mind addressing them...

File: lib/mongoid/touchable.rb
Comment (johnnyshields) [reply lib/mongoid/touchable.rb:43]: I think I got it from somewhere else in the code, I forget. I can make it single _ if you prefer.

The reason for _ prefix is similar to `@api private`, it denotes a method that really shouldn't be called externally.

File: lib/mongoid/touchable.rb
Comment (johnnyshields) [reply lib/mongoid/touchable.rb:49]: If I remember correctly, the reason for doing this is that, not only do we want to get the document updates (which are applied to root), we also want to mutate the objects in-memory to reflect the touch. Hence all the `write_attribute` calls. Does this make sense?

File: lib/mongoid/touchable.rb
Comment (johnnyshields) [reply lib/mongoid/touchable.rb:85]: Done.

File: lib/mongoid/touchable.rb
Comment (johnnyshields) [reply lib/mongoid/touchable.rb]: It is unnecessary, great catch. I've removed it and added a comment instead.

File: lib/mongoid/touchable.rb
Comment (johnnyshields) [reply lib/mongoid/touchable.rb]: I've gone ahead and used single _

File: lib/mongoid/touchable.rb
Comment (Neilshweky) [reply lib/mongoid/touchable.rb]: you can do all that without calling `__extract_touches_from_atomic_sets` on each "generation", no?

File: lib/mongoid/touchable.rb
Comment (Neilshweky) [reply lib/mongoid/touchable.rb]: okay cool

File: lib/mongoid/touchable.rb
Comment (Neilshweky) [reply lib/mongoid/touchable.rb:27]: yeah for sure... That one should be a lot more easily rebased than this one will... I wouldn't wish such a fate on you LMAO

File: spec/mongoid/touchable_spec.rb
Comment (johnnyshields) [reply spec/mongoid/touchable_spec.rb]: Originally I thought it wasn't needed, but why not? I've just added it.

File: lib/mongoid/touchable.rb
Comment (Neilshweky) [inline lib/mongoid/touchable.rb:148]:
```suggestion
              # If touch_field is nil, it will only touch the default field (updated_at).
```

**Comment by johnnyshields** [comment]
@Neilshweky all comments addressed.

File: lib/mongoid/touchable.rb
Comment (Neilshweky) [reply lib/mongoid/touchable.rb]: nitpick... will do a more full review when you address the rest of my comments

File: spec/mongoid/touchable_spec.rb
Comment (Neilshweky) [reply spec/mongoid/touchable_spec.rb:105]: yeah adding the default is not necessary on second thought.

File: lib/mongoid/touchable.rb
Comment (johnnyshields) [reply lib/mongoid/touchable.rb]: My brain is out of mana on this issue. I suggest we merge as-is and do the refactor as a follow-up.

You're probably right that it can be made more efficient in Ruby, however, the first thing this PR is trying to achieve is to make the database calls more efficient (which it achieves).

File: spec/mongoid/touchable_spec.rb
Comment (Neilshweky) [reply spec/mongoid/touchable_spec.rb]: Thanks!

**Comment by johnnyshields** [comment]
@Neilshweky I appreciate if you can take over this PR from here, if there are further changes needed. I have a busy few weeks ahead of me and won't be about to devote further attention for awhile.

File: lib/mongoid/touchable.rb
Comment (Neilshweky) [reply lib/mongoid/touchable.rb]: I'm not quite ready to approve this, if you don't mind just thinking out this last point with me.... I decided to see how difficult this would be to fix, and I discovered an interesting decision you made (I'm not sure if this was accidentally or purposefully), but I would like to bring it up.... 

I noticed that in the recursive call to `_gather_touch_updates` the field is not passed to the parent as well. Now, is this done on purpose, since the given field should only be touched on the deepest level, whereas on its ancestors the `updated_at` field should be touched? 

If this is true, then it makes sense to call `_extract_touches_from_atomic_sets` on each level.

Comment (Neilshweky) [reply lib/mongoid/touchable.rb]: However I noticed one other thing which is _not_ taken into account, and I would be okay with leaving that as a followup, but MONGOID-5508 is not taken into account here. The recursive call here would have to check the association for a touch field and pass it in.

Comment (Neilshweky) [reply lib/mongoid/touchable.rb]: Verify the above comments for me and then I should be okay to approve.

**Comment by Neilshweky** [comment]
@comandeo I am pretty much done with my review, if you wouldn't mind taking a look here.

File: lib/mongoid/touchable.rb
Comment (johnnyshields) [reply lib/mongoid/touchable.rb]: > Is this done on purpose, since the given field should only be touched on the deepest level, whereas on its ancestors the updated_at field should be touched.

The intent is that `child.touch(:field_a)` would touch `:field_a` and `:updated` on the child, and then *only* touch `:updated_at` (not `:field_a`) on the parent. Supposing I also have:

```
class Child
  embedded_in :parent, touch: :field_b
```

Then I need to touch `:field_b` and `:updated_at` on the parent, regardless of which field I touch on the child.

Does that make sense?

Comment (johnnyshields) [reply lib/mongoid/touchable.rb]: As we've noted elsewhere, there are more refactors needed especially for persistence-related auto-touching, so I expect we'll be refactoring all this again anyway.

Comment (Neilshweky) [reply lib/mongoid/touchable.rb]: yes it does make sense.... My latter point was that this: 
>  Then I need to touch :field_b and :updated_at on the parent, regardless of which field I touch on the child.

is not being accomplished in this PR. I'm okay with that, and I'm willing for that to be done in MONGOID-5508.

**Comment by Neilshweky** [comment]
Good job @johnnyshields, this really cleans up this part of the code.

**Review by comandeo — APPROVED** [review]
Great job! :clap:

**Comment by Neilshweky** [comment]
Open questions/todos:

- Should this go into 8.1/9.0?

Let's do 9.0 since the repairing of `touch: false` on embedded associations can be a BC break.

- What is default behavior of touch option for embedded associations?

Does this PR change the default touch behavior for embedded associations?

- Since we have decided to put this into 9.0 there needs to be release notes stating that `touch: false` for embedded associations has been repaired... @johnnyshields can you handle that? You can even open a separate PR if you like...

**Comment by johnnyshields** [comment]
@Neilshweky agreed it should be 9.0.

Are we also going to change the default value to touch: true for embedded? That will have some bearing on the release notes.

**Comment by Neilshweky** [comment]
I would love to change the default of touch to true for embedded associations, and I will write up a proposal for it in a ticket (MONGOID-5016?)... Assume that we will be changing the default to true, so that (2) will not change... Are there any backwards breaking changes in this PR? Anything that we would need to put in the release notes?

**Comment by johnnyshields** [comment]
There are no backwards breaking changes. The release notes for the touch: true could have a line like:

`In addition, there have been some refactor of the #touch method to generate fewer queries, and edge cases where touch: false was not working previously have been standardized.`

Important note `touch: true` would only be the new default for `embedded_in` relations (i.e. touch child causes touch on parent). Referenced relations need to stay as `touch: false`.

There's also some ticket to remove `touch: false` support on `embedded_in`, but I think that's a bridge too far for Mongoid 9.0 in terms of preserving behavior. Should revisit that later.

**Comment by Neilshweky** [comment]
@johnnyshields take a look at the last commit I made here. Turns out in one of the tests that used a `meth`, the `shared_examples` wasn't using it. I changed that and fixed the test (there's a problem in MONGOID-5504 that I will have to fix)... I also explicitly specified the `touch` option on all models so that when we flip the default, the tests won't fail... I'll add new tests when I do MONGOID-5016

**Comment by johnnyshields** [comment]
Ah yep, great catch! Last commit is good.

**Comment by Neilshweky** [comment]
In the interest of keeping my head on straight, here are the TODOs and followup work: 
- write 9.0 release notes @johnnyshields is handling this (if necessary)
- Followup: flip the default of the `touch` option on embedded associations to true MONGOID-5016
- Followup: verify that the `touch: :field` functionality works correctly MONGOID-5508

**Comment by Neilshweky** [comment]
> There are no backwards breaking changes. The release notes for the touch: true could have a line like:

I think this might be a good idea but we can do this as part of MONGOID-5016
> `In addition, there have been some refactor of the #touch method to generate fewer queries, and edge cases where touch: false was not working previously have been corrected.`
> 
> Important note - `touch: true` would only be the new default for `embedded_in` relations (i.e. touch child causes touch on parent). Referenced relations need to stay as `touch: false`.
 
This is MONGOID-5016 which I repurposed for switching the default
> There's also some ticket to remove `touch: false` support on `embedded_in`, but I think that's a bridge too far for Mongoid 9.0 in terms of preserving behavior. Should revisit that later.

**Comment by Neilshweky** [comment]
@johnnyshields even though there are no backwards breaking changes, the `touch: false` option has been fixed and I think that warrants a release note.

File: docs/release-notes/mongoid-9.0.txt
Comment (Neilshweky) [inline docs/release-notes/mongoid-9.0.txt:107]:
```suggestion
In addition, the ``#touch`` method has been optimized to perform one
```

**Review by Neilshweky — APPROVED** [review]

**Comment by johnnyshields** [comment]
@Neilshweky your spec change adding `touch: false` seems to have broken a few specs. Suggest to revert it and fix it in the next PR.

**Comment by Neilshweky** [comment]
I will fix it and merge when its done

**Pull request merged**
